File: connectors/clients/julia/src/vcl.jl
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# VeriSimDB Julia Client — VCL (VeriSimDB Query Language) operations.
#
# VCL is VeriSimDB's native query language for multi-modal queries that span
# graph traversals, vector similarity, spatial filters, and temporal constraints
# in a single statement. This file provides execution and explain functions.

"""
    execute_vcl(client, query; params=Dict()) -> VclResult

Execute a VCL query and return the result set.

VCL queries can combine modalities — for example:
```
FIND octads WHERE vector_similar(\$embedding, 0.8)
  AND spatial_within(51.5, -0.1, 10km)
  AND graph_connected("category:science", depth: 2)
```

# Arguments
- `client::Client` — The authenticated client.
- `query::String` — The VCL query string.

# Keyword Arguments
- `params::Dict{String,String}` — Named parameters for parameterised queries.

# Returns
A `VclResult` containing columns, rows, count, and execution time.
"""
function execute_vcl(
    client::Client,
    query::String;
    params::Dict{String,String}=Dict{String,String}()
)::VclResult
    body = Dict("query" => query, "params" => params)
    resp = do_post(client, "/api/v1/vcl/execute", body)
    return parse_response(VclResult, resp)
end

"""
    explain_vcl(client, query; params=Dict()) -> VclExplanation

Return the query execution plan for a VCL statement without running it.
Useful for debugging and optimising queries.

# Arguments
- `client::Client` — The authenticated client.
- `query::String` — The VCL query string.

# Keyword Arguments
- `params::Dict{String,String}` — Named parameters.

# Returns
A `VclExplanation` containing the plan, estimated cost, and warnings.
"""
function explain_vcl(
    client::Client,
    query::String;
    params::Dict{String,String}=Dict{String,String}()
)::VclExplanation
    body = Dict("query" => query, "params" => params)
    resp = do_post(client, "/api/v1/vcl/explain", body)
    return parse_response(VclExplanation, resp)
end

File: connectors/clients/rust/src/vcl.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

//! VeriSim Query Language (VCL) execution.
//!
//! VCL is VeriSimDB's native query language, supporting SQL-like syntax extended
//! with multi-modal operations (vector similarity, graph traversal, spatial
//! predicates, drift thresholds, etc.). This module provides methods to execute
//! VCL statements and retrieve explain / query plans.

use serde::{Deserialize, Serialize};

use crate::client::VeriSimClient;
use crate::error::Result;

/// Response from a VCL query execution or explain request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VclResponse {
    /// Whether the query executed successfully.
    pub success: bool,
    /// The type of VCL statement ("SELECT", "INSERT", "UPDATE", "DELETE", "EXPLAIN", etc.).
    pub statement_type: String,
    /// Number of rows affected or returned.
    pub row_count: usize,
    /// The result data (rows for SELECT, affected IDs for mutations, plan for EXPLAIN).
    pub data: serde_json::Value,
    /// Optional human-readable message (warnings, notices, etc.).
    pub message: Option<String>,
}

/// Internal request body for VCL execution.
#[derive(Debug, Serialize)]
struct VclRequest {
    query: String,
}

impl VeriSimClient {
    /// Execute a VCL statement against the VeriSimDB instance.
    ///
    /// Supports SELECT, INSERT, UPDATE, DELETE, and VeriSimDB-specific
    /// statements like `DRIFT CHECK`, `NORMALIZE`, and `FEDERATE`.
    ///
    /// # Arguments
    ///
    /// * `query` — The VCL statement string.
    ///
    /// # Errors
    ///
    /// Returns [`VeriSimError::Server`] if the query has syntax errors or
    /// the server rejects it for semantic reasons.
    pub async fn execute_vcl(&self, query: &str) -> Result<VclResponse> {
        let body = VclRequest {
            query: query.to_owned(),
        };
        self.post("/api/v1/vcl/execute", &body).await
    }

    /// Request an explain / query plan for a VCL statement without executing it.
    ///
    /// Useful for understanding which modalities, indices, and federation peers
    /// would be involved in a query.
    ///
    /// # Arguments
    ///
    /// * `query` — The VCL statement string to explain.
    pub async fn explain_vcl(&self, query: &str) -> Result<VclResponse> {
        let body = VclRequest {
            query: query.to_owned(),
        };
        self.post("/api/v1/vcl/explain", &body).await
    }
}
